overlord/devicestate/handlers_test.go

```diff
@@ -424,3 +429,129 @@ func (s *deviceMgrSuite) TestDoPrepareRemodeling(c *C) {
 	_, ok = devicestate.CachedRemodelCtx(chg)
 	c.Check(ok, Equals, false)
 }
+
+type preseedBaseSuite struct {
+	deviceMgrBaseSuite
+	restorePreseedMode func()
+	cmdUmount          *testutil.MockCmd
+}
+
+func (s *preseedBaseSuite) SetUpTest(c *C, preseed bool) {
+	s.restorePreseedMode = release.MockPreseedMode(func() bool { return preseed })
+	// preseed mode helper needs to be mocked before setting up
+	// deviceMgrBaseSuite due to device Manager init.
+	s.deviceMgrBaseSuite.SetUpTest(c)
+
+	s.cmdUmount = testutil.MockCommand(c, "umount", "")
+
+	st := s.state
+	st.Lock()
+	defer st.Unlock()
+
+	si := &snap.SideInfo{RealName: "test-snap", Revision: snap.R(3), SnapID: "test-snap-id"}
+	snaptest.MockSnap(c, `name: test-snap
+version: 1.0
+apps:
+ srv:
+  command: bin/service
+  daemon: simple
+`, si)
+
+	snapstate.Set(st, "test-snap", &snapstate.SnapState{
+		Active:   true,
+		Sequence: []*snap.SideInfo{si},
+		Current:  si.Revision,
+		SnapType: "app",
+	})
+}
+
+func (s *preseedBaseSuite) TearDownTest(c *C) {
+	s.deviceMgrBaseSuite.TearDownTest(c)
+	s.cmdUmount.Restore()
+}
+
+type preseedModeSuite struct {
+	preseedBaseSuite
+}
+
+var _ = Suite(&preseedModeSuite{})
+
+func (s *preseedModeSuite) SetUpTest(c *C) {
+	s.preseedBaseSuite.SetUpTest(c, true)
+}
+
+func (s *preseedModeSuite) TearDownTest(c *C) {
+	s.preseedBaseSuite.TearDownTest(c)
+}
+
+func (s *preseedModeSuite) TestDoMarkPreseeded(c *C) {
+	st := s.state
+	st.Lock()
+	defer st.Unlock()
+
+	chg := st.NewChange("firstboot seeding", "...")
+	t := st.NewTask("mark-preseeded", "...")
+	chg.AddTask(t)
+
+	st.Unlock()
+	s.se.Ensure()
+	s.se.Wait()
+	st.Lock()
+
+	// mark-preseeded task is left in Doing, meaning it will be re-executed
+	// after restart in normal (not preseeding) mode.
+	c.Check(t.Status(), Equals, state.DoingStatus)
+
+	var preseeded bool
+	c.Check(t.Get("preseeded", &preseeded), IsNil)
+	c.Check(preseeded, Equals, true)
+
+	// re-trying mark-preseeded task has no effect
```

Comment:
shouldn't we manually check that the status of `t` is still `Doing` after this again to prove that it has no effect?

Reply:
Added.